pybpodapi/bpod/emulator/emulator.py

```diff
@@ -0,0 +1,715 @@
+                        self._n_current_events += 1
+                    if virtual_current_events and \
+                            virtual_current_events[0] == \
+                            sma_global_counters.attached_events[i]:
+                        self._global_counter_counts[i] += 1
+            # Evaluate condition transitions
+            for i in range(self.hardware.n_conditions):
+                sma_conditions = self._state_machine.conditions
+                condition_event_name = CONDITION_EVENT_NAME_PREFIX + str(i + 1)
+                condition_event_code = \
+                    self.hardware.channels.event_names.index(
+                        condition_event_name)
+                target_state = None
+                for event_code, destination_state in sma_conditions.matrix[
+                        self._current_state]:
+                    if event_code == condition_event_code:
+                        target_state = destination_state
+                        break
+                if target_state is not None:
+                    if target_state != self._current_state:
+                        this_channel = sma_conditions.channels[i]
+                        hw_state = self._state.input[this_channel]
+                        if hw_state == sma_conditions.values[i]:
+                            virtual_current_events.append(
+                                self._CONDITION_OFFSET + i)
+                            self._n_current_events += 1
+            # Evaluate state timer transitions
+            time_in_state = self.current_time - self._state_start_time
+            state_timer = self._state_machine.state_timers[self._current_state]
+            if (time_in_state > state_timer) and \
+                    self._meaningful_timer[self._current_state]:
+                hardware_state_timer_position = \
+                    self.hardware_event_types.index(UART_CODE)
+                virtual_current_events.append(
+                    hardware_state_timer_position)
+                self._n_current_events += 1
+            if virtual_current_events:
+                dominant_event = virtual_current_events[0]
+                if dominant_event > 0:
+                    opcode = 1
+                    data = virtual_current_events
+                    total_n_events = self._n_events + self._n_current_events
+                    self._timestamps[
+                        self._n_events:total_n_events + 1] = \
+                        [self.current_time - self._matrix_start_time] * \
+                        self._n_current_events
+                    self._n_events = total_n_events
+        else:
+            opcode = 2
+            data = self._softcode
+            self._softcode = 0
+        return opcode, data
+
+    def mirror_state(self, state):
+        if state is None:
+            self._state.reset_input()
+            self._state.reset_output()
+            # BpodSystem.RefreshGUI;
```

Comment:
This needs to be a call to the gui-emulator-plugin if it's enabled, correct?. If so, then maybe it's worth adding a `#TODO: ....`